File: casa-3.2.1/wcslib/Fortran/test/ttab1.f
*=======================================================================
*
*   WCSLIB 4.3 - an implementation of the FITS WCS standard.
*   Copyright (C) 1995-2007, Mark Calabretta
*
*   This file is part of WCSLIB.
*
*   WCSLIB is free software: you can redistribute it and/or modify
*   it under the terms of the GNU Lesser General Public License as
*   published by the Free Software Foundation, either version 3 of
*   the License, or (at your option) any later version.
*
*   WCSLIB is distributed in the hope that it will be useful, but
*   WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU Lesser General Public License for more details.
*
*   You should have received a copy of the GNU Lesser General Public
*   License along with WCSLIB.  If not, see http://www.gnu.org/licenses.
*
*   Correspondence concerning WCSLIB may be directed to:
*      Internet email: mcalabre@atnf.csiro.au
*      Postal address: Dr. Mark Calabretta
*                      Australia Telescope National Facility, CSIRO
*                      PO Box 76
*                      Epping NSW 1710
*                      AUSTRALIA
*
*   Author: Mark Calabretta, Australia Telescope National Facility
*   http://www.atnf.csiro.au/~mcalabre/index.html
*   $Id: ttab1.f,v 4.3 2007/12/27 05:42:52 cal103 Exp $
*=======================================================================

      PROGRAM TTAB1
*-----------------------------------------------------------------------
*
*   TTAB1 tests the -TAB routines for closure.
*
*-----------------------------------------------------------------------

      DOUBLE PRECISION TOL
      PARAMETER (TOL = 1D-8)

      INTEGER   K1, K2, M
      PARAMETER (M = 2, K1 = 32, K2 = 16)

      INTEGER   K(2), MAP(2)
      DOUBLE PRECISION CRVAL(2)
      DATA K     /K1, K2/
      DATA MAP   /0, 1/
      DATA CRVAL /1D0, -1D0/

      INTEGER   I, IK, IK1, IK2, IM, J, N, STAT0(128), STAT1(128),
     :          STATUS
      DOUBLE PRECISION CRPIX4, EPSILON, RESID, RESIDMAX, TIME(12),
     :          WORLD(M,11,11), XT0(12), XT1(12), X0(M,11,11),
     :          X1(M,11,11), Z

      INCLUDE 'tab.inc'
      INTEGER   TAB(TABLEN)
*-----------------------------------------------------------------------

      WRITE (*, 10) TOL
 10   FORMAT ('Testing closure of WCSLIB tabular coordinate routines ',
     :        '(ttab1.f)',/,
     :        '------------------------------------------------------',
     :        '---------',//,
     :        'Reporting tolerance',1PG8.1,'.')

*     First a 1-dimensional table from Sect. 6.2.3 of Paper III.
      WRITE (*, '(/,A)') 'One-dimensional test:'
      STATUS = TABPUT (TAB, TAB_FLAG, -1, 0, 0)
      STATUS = TABINI (M, K, TAB)
      IF (STATUS.NE.0) THEN
         WRITE (*, 20) STATUS
 20      FORMAT ('TABINI ERROR',I2,'.')
         GO TO 999
      END IF

      STATUS = TABPUT (TAB, TAB_M,       1, 0, 0)
      STATUS = TABPUT (TAB, TAB_K,       8, 1, 0)
      STATUS = TABPUT (TAB, TAB_MAP,     0, 1, 0)
      STATUS = TABPUT (TAB, TAB_CRVAL, 0D0, 1, 0)

      STATUS = TABPUT (TAB, TAB_INDEX, 0D0, 1, 1)
      STATUS = TABPUT (TAB, TAB_INDEX, 1D0, 1, 2)
      STATUS = TABPUT (TAB, TAB_INDEX, 1D0, 1, 3)
      STATUS = TABPUT (TAB, TAB_INDEX, 2D0, 1, 4)
      STATUS = TABPUT (TAB, TAB_INDEX, 2D0, 1, 5)
      STATUS = TABPUT (TAB, TAB_INDEX, 3D0, 1, 6)
      STATUS = TABPUT (TAB, TAB_INDEX, 3D0, 1, 7)
      STATUS = TABPUT (TAB, TAB_INDEX, 4D0, 1, 8)

      STATUS = TABPUT (TAB, TAB_COORD, 1997.84512D0, 1, 0)
      STATUS = TABPUT (TAB, TAB_COORD, 1997.84631D0, 2, 0)
      STATUS = TABPUT (TAB, TAB_COORD, 1993.28451D0, 3, 0)
      STATUS = TABPUT (TAB, TAB_COORD, 1993.28456D0, 4, 0)
      STATUS = TABPUT (TAB, TAB_COORD, 2001.59234D0, 5, 0)
      STATUS = TABPUT (TAB, TAB_COORD, 2001.59239D0, 6, 0)
      STATUS = TABPUT (TAB, TAB_COORD, 2002.18265D0, 7, 0)
      STATUS = TABPUT (TAB, TAB_COORD, 2002.18301D0, 8, 0)

      EPSILON = 1D-3
      CRPIX4  = 0.5
      XT0(1)  = 0.5 + EPSILON - CRPIX4
      XT0(2)  = 1.0           - CRPIX4
      XT0(3)  = 1.5 - EPSILON - CRPIX4
      XT0(4)  = 1.5 + EPSILON - CRPIX4
      XT0(5)  = 2.0           - CRPIX4
      XT0(6)  = 2.5 - EPSILON - CRPIX4
      XT0(7)  = 2.5 + EPSILON - CRPIX4
      XT0(8)  = 3.0           - CRPIX4
      XT0(9)  = 3.5 - EPSILON - CRPIX4
      XT0(10) = 3.5 + EPSILON - CRPIX4
      XT0(11) = 4.0           - CRPIX4
      XT0(12) = 4.5 - EPSILON - CRPIX4

      STATUS = TABX2S (TAB, 12, 1, XT0, TIME, STAT0)
      IF (STATUS.NE.0) THEN
         WRITE (*, 30) STATUS
 30      FORMAT ('TABX2S ERROR',I2,'.')
      END IF

      STATUS = TABS2X (TAB, 12, 1, TIME, XT1, STAT1)
      IF (STATUS.NE.0) THEN
         WRITE (*, 40) STATUS
 40      FORMAT ('TABX2S ERROR',I2,'.')
      END IF

      WRITE (*, 50)
 50   FORMAT ('    x   ->   time   ->   x')
      DO 70 I = 1, 12
         WRITE (*, 60) XT0(I), TIME(I), XT1(I)
 60      FORMAT (F8.5,F12.5,F9.5)
 70   CONTINUE
      WRITE (*, '(/)')

*     Test closure.
      RESIDMAX = 0D0
      DO 110 I = 1, 12
         IF (STAT0(I).NE.0) THEN
            WRITE (*, 80) XT0(I), STAT0(I)
 80         FORMAT ('   TABX2S: X =',F7.1,', STAT =',I2)
            GO TO 110
         END IF

         IF (STAT1(I).NE.0) THEN
            WRITE (*, 90) TIME(I), STAT1(I)
 90         FORMAT ('   TABS2X: T =',F7.1,', STAT =',I2)
            GO TO 110
         END IF

         RESID = ABS(XT1(I) - XT0(I))
         IF (RESID.GT.RESIDMAX) RESIDMAX = RESID

         IF (RESID.GT.TOL) THEN
            WRITE (*, 100) XT0(I), TIME(I), XT1(I)
 100        FORMAT ('   Closure error:',/,'      X = ',F20.15,/,
     :              '   -> T = ',F20.15,/,'   -> X = ',F20.15)
         END IF
 110  CONTINUE

      STATUS = TABFREE (TAB)


*     Now a 2-dimensional table.
      WRITE (*, '(A)') 'Two-dimensional test:'
      STATUS = TABPUT (TAB, TAB_FLAG, -1, 0, 0)
      STATUS = TABINI (M, K, TAB)
      IF (STATUS.NE.0) THEN
         WRITE (*, 120) STATUS
 120     FORMAT ('TABINI ERROR',I2,'.')
         GO TO 999
      END IF

      STATUS = TABPUT (TAB, TAB_M, M, 0, 0)
      DO 140 IM = 1, M
         STATUS = TABPUT (TAB, TAB_K,     K(IM),     IM, 0)
         STATUS = TABPUT (TAB, TAB_MAP,   MAP(IM),   IM, 0)
         STATUS = TABPUT (TAB, TAB_CRVAL, CRVAL(IM), IM, 0)

         DO 130 IK = 1, K(IM)
            STATUS = TABPUT (TAB, TAB_INDEX, DBLE(IK-1), IM, IK)
 130     CONTINUE
 140  CONTINUE

      N = 0
      Z = 1D0 / ((K1-1) * (K2-1))
      DO 160 IK2 = 0, K2-1
         DO 150 IK1 = 0, K1-1
            N = N + 1
            STATUS = TABPUT (TAB, TAB_COORD,  3D0*IK1*IK2*Z, N, 0)
            N = N + 1
            STATUS = TABPUT (TAB, TAB_COORD, -1D0*(K1-IK1-1)*IK2*Z +
     :                       0.01*IK1, N, 0)
 150     CONTINUE
 160  CONTINUE

      DO 180 I = 1, 11
         DO 170 J = 1, 11
            X0(1,J,I) = (J-1)*(K1-1)/10D0 - CRVAL(1)
            X0(2,J,I) = (I-1)*(K2-1)/10D0 - CRVAL(2)
 170     CONTINUE
 180  CONTINUE

      STATUS = TABX2S (TAB, 121, 2, X0, WORLD, STAT0)
      IF (STATUS.NE.0) THEN
         WRITE (*, 190) STATUS
 190     FORMAT ('TABX2S ERROR',I2,'.')
      END IF

      STATUS = TABS2X (TAB, 121, 2, WORLD, X1, STAT1)
      IF (STATUS.NE.0) THEN
         WRITE (*, 200) STATUS
 200     FORMAT ('TABX2S ERROR',I2,'.')
      END IF

*     Test closure.
      N = 0
      RESIDMAX = 0D0
      DO 260 I = 1, 11
         DO 250 J = 1, 11
            N = N + 1
            IF (STAT0(N).NE.0) THEN
               WRITE (*, 210) X0(1,J,I), X0(2,J,I), STAT0(N)
 210           FORMAT ('   TABX2S: X = (',F6.1,',',F6.1,'), STAT =',I2)
               GO TO 250
            END IF

            IF (STAT1(N).NE.0) THEN
               WRITE (*, 220) WORLD(1,J,I), WORLD(2,J,I), STAT1(N)
 220           FORMAT ('   TABS2X: S = (',F6.1,',',F6.1,'), STAT =',I2)
               GO TO 250
            END IF

            DO 240 IM = 1, M
               RESID = ABS(X1(IM,J,I) - X0(IM,J,I))
               IF (RESID.GT.RESIDMAX) RESIDMAX = RESID

               IF (RESID.GT.TOL) THEN
                  WRITE (*, 230) X0(1,J,I), X0(2,J,I), WORLD(1,J,I),
     :                           WORLD(2,J,I), X1(1,J,I), X1(2,J,I)
 230              FORMAT ('   Closure error:',/,
     :                    '      X = (',F20.15,',',F20.15,')',/,
     :                    '   -> W = (',F20.15,',',F20.15,')',/,
     :                    '   -> X = (',F20.15,',',F20.15,')')
                  GO TO 250
               END IF
 240        CONTINUE
 250     CONTINUE
 260  CONTINUE

      WRITE (*, 270) RESIDMAX
 270  FORMAT ('Maximum closure residual =',1PE19.12)


 999  STATUS = TABFREE (TAB)

      END

File: casa-3.2.1/wcslib/C/test/ttab1.c
/*============================================================================

    WCSLIB 4.3 - an implementation of the FITS WCS standard.
    Copyright (C) 1995-2007, Mark Calabretta

    This file is part of WCSLIB.

    WCSLIB is free software: you can redistribute it and/or modify it under
    the terms of the GNU Lesser General Public License as published by the
    Free Software Foundation, either version 3 of the License, or (at your
    option) any later version.

    WCSLIB is distributed in the hope that it will be useful, but WITHOUT ANY
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
    more details.

    You should have received a copy of the GNU Lesser General Public License
    along with WCSLIB.  If not, see <http://www.gnu.org/licenses/>.

    Correspondence concerning WCSLIB may be directed to:
       Internet email: mcalabre@atnf.csiro.au
       Postal address: Dr. Mark Calabretta
                       Australia Telescope National Facility, CSIRO
                       PO Box 76
                       Epping NSW 1710
                       AUSTRALIA

    Author: Mark Calabretta, Australia Telescope National Facility
    http://www.atnf.csiro.au/~mcalabre/index.html
    $Id: ttab1.c,v 4.3 2007/12/27 05:35:51 cal103 Exp $
*=============================================================================
*
*   ttab1 tests the -TAB routines for closure.
*
*---------------------------------------------------------------------------*/

#include <math.h>
#include <stdio.h>

#include <tab.h>

/* Reporting tolerance. */
const double tol = 1.0e-8;

#define K1 32
#define K2 16

int main()

{
   const int M = 2;
   const int K[] = {K1, K2};
   const int map[] = {0, 1};
   const double crval[] = {1.0, -1.0};

   int i, j, k, k1, k2, m, n, stat0[128], stat1[128], status;
   double crpix4, epsilon, resid, residmax, time[12], world[11][11][2],
          xt0[12], xt1[12], x0[11][11][2], x1[11][11][2], z;
   struct tabprm tab;

   printf(
      "Testing closure of WCSLIB tabular coordinate routines (ttab1.c)\n"
      "---------------------------------------------------------------\n");

   /* List status return messages. */
   printf("\nList of tab status return values:\n");
   for (status = 1; status <= 5; status++) {
      printf("%4d: %s.\n", status, tab_errmsg[status]);
   }

   printf("\nReporting tolerance %5.1G.\n", tol);


   /* First a 1-dimensional table from Sect. 6.2.3 of Paper III. */
   printf("\nOne-dimensional test:\n");
   tab.flag = -1;
   if ((status = tabini(1, M, K, &tab))) {
      printf("tabini ERROR %d: %s.\n", status, tab_errmsg[status]);
      return 1;
   }

   tab.M = 1;
   tab.K[0] = 8;
   tab.map[0] = 0;
   tab.crval[0] = 0.0;

   tab.index[0][0] = 0.0;
   tab.index[0][1] = 1.0;
   tab.index[0][2] = 1.0;
   tab.index[0][3] = 2.0;
   tab.index[0][4] = 2.0;
   tab.index[0][5] = 3.0;
   tab.index[0][6] = 3.0;
   tab.index[0][7] = 4.0;

   tab.coord[0] = 1997.84512;
   tab.coord[1] = 1997.84631;
   tab.coord[2] = 1993.28451;
   tab.coord[3] = 1993.28456;
   tab.coord[4] = 2001.59234;
   tab.coord[5] = 2001.59239;
   tab.coord[6] = 2002.18265;
   tab.coord[7] = 2002.18301;

   epsilon = 1e-3;
   crpix4  = 0.5;
   xt0[0]  = 0.5 + epsilon - crpix4;
   xt0[1]  = 1.0           - crpix4;
   xt0[2]  = 1.5 - epsilon - crpix4;
   xt0[3]  = 1.5 + epsilon - crpix4;
   xt0[4]  = 2.0           - crpix4;
   xt0[5]  = 2.5 - epsilon - crpix4;
   xt0[6]  = 2.5 + epsilon - crpix4;
   xt0[7]  = 3.0           - crpix4;
   xt0[8]  = 3.5 - epsilon - crpix4;
   xt0[9]  = 3.5 + epsilon - crpix4;
   xt0[10] = 4.0           - crpix4;
   xt0[11] = 4.5 - epsilon - crpix4;

   if ((status = tabx2s(&tab, 12, 1, (double *)xt0, (double *)time, stat0))) {
      printf("tabx2s ERROR %d: %s.\n", status, tab_errmsg[status]);
   }

   if ((status = tabs2x(&tab, 12, 1, (double *)time, (double *)xt1, stat1))) {
      printf("tabx2s ERROR %d: %s.\n", status, tab_errmsg[status]);
   }

   printf("    x   ->   time   ->   x\n");
   for (i = 0; i < 12; i++) {
      printf("%8.5f%12.5f%9.5f\n", xt0[i], time[i], xt1[i]);
   }
   printf("\n");

   /* Test closure. */
   residmax = 0.0;
   for (i = 0; i < 12; i++) {
      if (stat0[i]) {
         printf("   tabx2s: x = %6.1f, stat = %d\n", xt0[i], stat0[i]);
         continue;
      }

      if (stat1[i]) {
         printf("   tabs2x: t = %6.1f, stat = %d\n", time[i], stat1[i]);
         continue;
      }

      resid = fabs(xt1[i] - xt0[i]);
      if (resid > residmax) residmax = resid;

      if (resid > tol) {
         printf("   Closure error:\n");
         printf("      x = %20.15f\n", xt0[i]);
         printf("   -> t = %20.15f\n", time[i]);
         printf("   -> x = %20.15f\n", xt1[i]);
      }
   }

   tabfree(&tab);


   /* Now a 2-dimensional table. */
   printf("Two-dimensional test:\n");
   tab.flag = -1;
   if ((status = tabini(1, M, K, &tab))) {
      printf("tabini ERROR %d: %s.\n", status, tab_errmsg[status]);
      return 1;
   }

   tab.M = M;
   for (m = 0; m < tab.M; m++) {
      tab.K[m] = K[m];
      tab.map[m] = map[m];
      tab.crval[m] = crval[m];

      for (k = 0; k < tab.K[m]; k++) {
         tab.index[m][k] = (double)k;
      }
   }

   n = 0;
   z = 1.0 / ((K1-1) * (K2-1));
   for (k2 = 0; k2 < K2; k2++) {
      for (k1 = 0; k1 < K1; k1++) {
         tab.coord[n++] =  3.0*k1*k2*z;
         tab.coord[n++] = -1.0*(K1-k1-1)*k2*z + 0.01*k1;
      }
   }

   for (i = 0; i < 11; i++) {
      for (j = 0; j < 11; j++) {
         x0[i][j][0] = j*(K1-1)/10.0 - crval[0];
         x0[i][j][1] = i*(K2-1)/10.0 - crval[1];
      }
   }

   if ((status = tabx2s(&tab, 121, 2, (double *)x0, (double *)world,
                        stat0))) {
      printf("tabx2s ERROR %d: %s.\n", status, tab_errmsg[status]);
   }

   if ((status = tabs2x(&tab, 121, 2, (double *)world, (double *)x1,
                        stat1))) {
      printf("tabx2s ERROR %d: %s.\n", status, tab_errmsg[status]);
   }

   /* Test closure. */
   n = 0;
   residmax = 0.0;
   for (i = 0; i < 11; i++) {
      for (j = 0; j < 11; j++, n++) {
         if (stat0[n]) {
            printf("   tabx2s: x = (%6.1f,%6.1f), stat = %d\n", x0[i][j][0],
               x0[i][j][1], stat0[n]);
            continue;
         }

         if (stat1[n]) {
            printf("   tabs2x: s = (%6.1f,%6.1f), stat = %d\n",
               world[i][j][0], world[i][j][1], stat1[n]);
            continue;
         }

         for (m = 0; m < M; m++) {
            resid = fabs(x1[i][j][m] - x0[i][j][m]);
            if (resid > residmax) residmax = resid;

            if (resid > tol) {
               printf("   Closure error:\n");
               printf("      x = (%20.15f,%20.15f)\n", x0[i][j][0],
                 x0[i][j][1]);
               printf("   -> w = (%20.15f,%20.15f)\n", world[i][j][0],
                 world[i][j][1]);
               printf("   -> x = (%20.15f,%20.15f)\n", x1[i][j][0],
                 x1[i][j][1]);

               break;
            }
         }
      }
   }

   printf("Maximum closure residual = %.12E\n", residmax);

   tabfree(&tab);

   return 0;
}
